Skip adding -lstatic to libtest and libstd on Darwin

gferon opened the pull request:
This is one way to fix #1573, which I have a feeling can be wrong. The binaries produced with this patch contain all required symbols and are working fine on macOS.

gferon commented:
Looks like the CI failure is unrelated to this change 🤔

illicitonion reviewed (changes requested):
It seems a little weird that we'd express "don't dynamically link libstd" by saying "don't statically link libstd", but, if it seems to work (which CI shows it does), that seems better than the status quo...

Would it be possible to get a couple of tests test showing the desired end-effect here?

1. Our existing binaries all successfully run on CI at the moment - do you have an example that fails without this patch?
2. Could we add a `sh_test` which runs `otool -L` on a produced `rust_binary` and shows that it doesn't dynamically link the libraries we don't expect to be dynamically linked?

gferon commented:
You're right, this would be much better if the issue was documented and a test would prevent any regression. I'll remove the patch on our codebase and try to come up with a minimal non-working example ASAP!

gferon commented:
I was able to reproduce the issue with a very simple dependency setup, full workspace available here: https://github.com/gferon/bazel-workspace-rules-rust-pr-1620

The problem is visible when you have a `rust_library` > `cc_library` > `rust_binary`. I don't have any idea why this is a problem, maybe @scentini has some insights?

```starlark
rust_library(
    name = "libtool-rs",
    srcs = ["lib.rs"],
    edition = "2021",
)

cc_library(
    name = "api",
    srcs = ["api.c"],
    deps = [":libtool-rs"],
)

rust_binary(
    name = "app",
    srcs = ["main.rs"],
    edition = "2021",
    deps = [":api"],
)
```

gferon commented:
@illicitonion since I'm currently working on upstreaming all of our custom patches, I unburried this one from the past. I added a test to demonstrate the issue, and I can confirm the test fails without the patch and passes otherwise.

> Could we add a sh_test which runs otool -L on a produced rust_binary and shows that it doesn't dynamically link the libraries we don't expect to be dynamically linked?

I added a `sh_test` that just runs the test binary, which has the same effect.

illicitonion commented on test/portable_link_flags_darwin/test.sh:
```diff
@@ -0,0 +1,4 @@
+#!/bin/sh
+set -e
+
+$1
```
Could you instead run `otool -L "$1"` and verify (by grep or whatever) that it doesn't contain "libstd" or "libtest"?

I want us to be sure that we're actually statically linking stuff, rather than setting up rpaths or whatever so that in the bazel-bin tree the binary _happens_ to work but if you moved it somewhere else or copied it to another machine it would fail to.

You can use `target_compatible_with` to make sure this test only runs on darwin where `otool` will be present.

illicitonion reviewed (commented):
This looks great, thanks so much! One small comment about the test, then let's get this merged!

illicitonion commented:
Thanks!
